File: cpp/july/day_22_Binary Tree Zigzag Level Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void traverse(TreeNode *root, vector<vector<int>> &res, int level, int &max_level){
        if(root){
            if(level == max_level && res.size() == level){
                res.push_back(vector<int>());
                max_level++;   
            }
            res[level].push_back(root->val);
            
            traverse(root->left, res, level+1, max_level);
            traverse(root->right, res, level+1, max_level);
        }
        
    }
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if(!root){
            return res;
        }
        int max_level = 0, i;
        traverse(root, res, 0, max_level);
        for(i = 1; i < res.size(); i+=2){
            reverse(res[i].begin(), res[i].end());
        }
        return res;
    }
};

File: python3/july/day_22_Binary Tree Zigzag Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def traverse(self, root, level):
        if(root):
            if(level == self.max_level and len(self.res) == level):
                self.res.append([])
                self.max_level += 1
            
            self.res[level].append(root.val)

            self.traverse(root.left, level+1)
            self.traverse(root.right, level+1)
    
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        self.res = []
        if(not root):
            return self.res
        
        self.max_level = 0
        self.traverse(root, 0)
        for i in range(1, len(self.res), 2):
            self.res[i] = self.res[i][::-1]

        return self.res
